Fix/15322 upgrade Stories-Android to android 11 api level 30

ravishanker opened the pull request:
Fixes [WordPress-Android#15339](https://github.com/wordpress-mobile/WordPress-Android/issues/15339)

Upgrade to Android 11 (API level 30)

Please see paqN3M-mr-p2 for details on mandatory requirement to upgrade to Android 11 by November 2021.  Also this paqN3M-n4-p2, for details on planned approach to upgrade

- upgrade targetSdkVersion and compileSdkVersion to 30
- fix and tweak a few files
- remove redundant permission WRITE_EXTERNAL_STORAGE

NOTE: Stories is first cab off the rank for upgrade as a dependency for WPAndroid upgrade

ravishanker committed:
Upgrade to 30

Upgraded the following to 30 from 28
targetSdkVersion,
compileSdkVersion

Made tweaks to a few files mostly guard against nulls

ravishanker committed:
Update AndroidManifest.xml

Remove redundant permission

ParaskP7 commented on build.gradle:
```diff
@@ -91,11 +91,11 @@ subprojects {
 
     minSdkVersion = 24
-    targetSdkVersion = 28
+    targetSdkVersion = 30
```
Suggestion (💡): Since the `buildToolVersion` is not needed and actually not used anymore, we can remove that as well. Afterwards, with some reordering this might look like:
```
    minSdkVersion = 24
    compileSdkVersion = 30
    targetSdkVersion = 30

```

ravishanker replied:
removed buildToolVersion. thanks

ParaskP7 commented on app/src/main/java/com/automattic/loop/StoryComposerActivity.kt:
```diff
@@ -78,11 +78,12 @@ class StoryComposerActivity : ComposeLoopFrameActivity(),
-                // if there are any external Uris in this list, copy them locally before trying to use them
-                processExternalUris(uriList)
+                val mediaUris = data.getStringArrayExtra(PhotoPickerActivity.EXTRA_MEDIA_URIS)
+                if (mediaUris != null) {
```
Suggestion (💡): Using `?.let { ... }` instead of an if not null check it might save us a line and make this a bit more idiomatic:

PS: Same applies to `setMediaUris(...)` on the `ComposeLoopFrameActivity.kt` class.

ravishanker replied:
done

ParaskP7 replied:
👋 @ravishanker !

What I meant was the `data.getStringArrayExtra(PhotoPickerActivity.EXTRA_MEDIA_URIS)?.let { ... }` equivalent, which will save a line, but please don't bother since this is totally minor, I just posted it as a nice-to-have.

😅

ParaskP7 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -5,7 +5,6 @@
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
-    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
Question (❓): Are we sure about that? Why was that there is the first place?

PS: Obviously, this is just an example `app` used to demonstrate that the `stories` functionality works as expected, prior to integrating that with `WordPress-Android`, but still, this is good to know, maybe this will be needed to be done on the main repo as well. 🤷

ravishanker replied:
Not sure why it was there.  It could be that because there's a READ so, WRITE was added by default.  It is definitely not needed here.  I've it on my list to check this again on main WPAndroid and test it and remove it.

ParaskP7 replied:
Thank you @ravishanker ! 🙏

mzorz replied:
👋  - this is there so others wanting to integrate the library in their projects can use the demo app as an example to be followed in their own apps, regarding what things are needed. In this sense, the `<uses-permission>` tag is not strictly needed for a demo app to be built or run, but it serves a purpose that goes beyond that. That specific tag is used by Google Play to filter apps through, so it will be needed if users of the library want to publish their apps there. For reference, here's the [documentation](https://developer.android.com/guide/topics/manifest/uses-permission-element).

Also:

> I've it on my list to check this again on main WPAndroid and test it and remove it

It won't affect WPAndroid, given this is only within the demo app's manifest.

Given the above reasons, please, can we get it back as it was @ravishanker ? 🙏 

As a side note, I believe this change is not related to the SDK change as advertised in the PR description. I would highly encourage unrelated changes to be either explained or be kept to a different PR when possible.

ParaskP7 reviewed (commented):
👋 @ravishanker  !

I have reviewed and tested this PR as per the instructions, it seems to work as expected, good job! 🌟 

I have left a couple of minor suggestions (💡) for you to consider. I am **NOT** going to approve this PR for now and until we make sure that `WordPress-Android` is ready as well, since it seems to me that the integration should happen at the same time. Correct me if I am wrong.

I also suggest we add @oguzkocer to the picture just to he is aware of this upcoming change.

Btw, have you tried to test this more thoroughly by integrating this branch with `WordPress-Android`? I honestly didn't go that far, but it seems wise if we do so before merging this PR.

From a conversation that I had with @oguzkocer a while ago (see [here](https://github.com/Automattic/stories-android/pull/702#issuecomment-861571309)) one can check out the branch in `libs/stories-android` folder of `WordPress-Android`. Then, if the integration works as well, we can have more confidence that there is nothing more that needs to be done for this upgrade. PS: Also, since we are upgrading two versions at once (from SDK `28` to `30`), I suggest we should be more careful with our testing here.

oguzkocer commented:
@ravishanker @ParaskP7 I am currently on support rotation, so I am not supposed to review any PRs 😅 Please ping me on Slack if you are blocked and I'll do my best to unblock you.

> Btw, have you tried to test this more thoroughly by integrating this branch with WordPress-Android?

@ParaskP7 is correct. Any stories-android library change should be tested in WPAndroid before merging the PR.

We also always ping @aforcier or @cameronvoell for a review when we make a change to stories-android library and get their approval before merge. (unless it's only a tooling change)

> Please see this P2 post for details on mandatory requirement to upgrade to Android 11 by November 2021.

@ravishanker We prefer not to link internal documents directly in the public repositories. There is a way to do this using a shortcode, you can see an example of how @AliSoftware does it [in this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/15319). I personally don't link any internal documents, not even as shortcodes, and prefer to provide/duplicate the necessary information in the public space for the benefit of non-a12s, however using shortcodes is widely accepted, so it's up to you.

ParaskP7 commented:
Thank you for your input @oguzkocer , much appreciated, happy support rotation! 🙇

AliSoftware commented:
> @ravishanker We prefer not to link internal documents directly in the public repositories. There is a way to do this using a shortcode

I've edited the PR description to use the shortcode instead.

@ravishanker if you want to avoid making that mistake in the future, I highly recommend installing our internal "GitHub Highlighter/Shorthander" tampermonkey script. You can find it by searching `github highlighter` in MGS.

ravishanker commented:
> 👋 @ravishanker !
> 
> I have reviewed and tested this PR as per the instructions, it seems to work as expected, good job! 🌟
> 
> I have left a couple of minor suggestions (💡) for you to consider. I am **NOT** going to approve this PR for now and until we make sure that `WordPress-Android` is ready as well, since it seems to me that the integration should happen at the same time. Correct me if I am wrong.
> 
> I also suggest we add @oguzkocer to the picture just to he is aware of this upcoming change.
> 
> Btw, have you tried to test this more thoroughly by integrating this branch with `WordPress-Android`? I honestly didn't go that far, but it seems wise if we do so before merging this PR.
> 
> From a conversation that I had with @oguzkocer a while ago (see [here](https://github.com/Automattic/stories-android/pull/702#issuecomment-861571309)) one can check out the branch in `libs/stories-android` folder of `WordPress-Android`. Then, if the integration works as well, we can have more confidence that there is nothing more that needs to be done for this upgrade. PS: Also, since we are upgrading two versions at once (from SDK `28` to `30`), I suggest we should be more careful with our testing here.

Thanks for the review @ParaskP7.  I started from WPAndroid, then moved on to stories-android.   I realized that, I need to fix all the dependencies starting with stories first.  So, yeah, I've tested with WPAndroid.

ravishanker committed:
Update for review

ParaskP7 commented:
👋 @ravishanker !

> Thanks for the review @ParaskP7. I started from WPAndroid, then moved on to stories-android. I realized that, I need to fix all the dependencies starting with stories first. So, yeah, I've tested with WPAndroid.

Great, thanks for confirming that you tested this branch with `WPAndroid`, nice job! 🌟 

PS: Then, I can approve this PR as nothing seems to be blocking. I however think that it might be better to merge than when `WPAndroid` is ready with the upgrade as well, wdyt?

mzorz commented:
👋 hello @ravishanker @ParaskP7 -  let me give this one a quick round of testing before we merge 👍

ParaskP7 commented:
👋 @mzorz !

> 👋 hello @ravishanker @ParaskP7 - let me give this one a quick round of testing before we merge 👍

💯 and many thanks! 🙇

mzorz commented on stories/src/main/java/com/wordpress/stories/compose/ComposeLoopFrameActivity.kt:
```diff
@@ -925,6 +917,15 @@ abstract class ComposeLoopFrameActivity : AppCompatActivity(), OnStoryFrameSelec
     }
 
+    private fun setMediaUris(intent: Intent) {
+        val mediaUris = intent.getStringArrayExtra(requestCodes.EXTRA_MEDIA_URIS)
```
Thank you for finding this opportunity to refactor 👍 

Nitpick: `val mediaUris` here can be removed and the expression simplified to `intent.getStringArrayExtra(requestCodes.EXTRA_MEDIA_URIS)?.let {...}`, given it's only used once

ravishanker replied:
done, thank you

mzorz reviewed (changes requested):
Thank you for taking on this task, @ravishanker !

I've smoke-tested both the demo app and WordPress-Android app with the stories-android library pointing to this branch, and using [this test branch on WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) (as some changes needed to be made in there as well).

- camera: taking pictures and recording videos worked as expected
- saving: both new pics and videos were composed successfully
- sourcing media: using the media picker also worked well (local media, WP media in the WP app, 3rd party app media, remote media).
- integration with Gutenberg (creating a post from zero, then editing, then editing an existing Post created on another device) seemed to work fine.

All of this using a Google Pixel 4a.

I also verified most of the changes in this PR are due to new build errors that appear from changing the `targetSdk` value to 30, so that's fine.

**IMPORTANT**: It is worth noting that when compiling the WordPress-Android app, the stories library will take the `targetSdk` and `compileSdk` from the parent project so, what this really means is that when building WPAndroid if we don't perform the changes [as expressed in this diff](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) then we'd only have tested the **_code_** changes in this PR but **not** the actual SDK change which this PR's main aim is supposed to be cc @ravishanker . I'd only ask that we make a more thorough list of what's been tested and how next time 👍 .

So, this **needs to be re-tested** once the parent app (WPAndroid) is ready for the change to API level 30, before we update the commit hash of submodule `/libs/stories-android` in the parent repository.

That said, let's please revert the `<uses-permission>` tag removal, and we can merge afterwards :shipit: 

Thank you!

ravishanker committed:
Update for review

Revert WRITE_EXTERNAL_STORAGE permission

ravishanker commented:
> Thank you for taking on this task, @ravishanker !
> 
> I've smoke-tested both the demo app and WordPress-Android app with the stories-android library pointing to this branch, and using [this test branch on WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) (as some changes needed to be made in there as well).
> 
>     * camera: taking pictures and recording videos worked as expected
> 
>     * saving: both new pics and videos were composed successfully
> 
>     * sourcing media: using the media picker also worked well (local media, WP media in the WP app, 3rd party app media, remote media).
> 
>     * integration with Gutenberg (creating a post from zero, then editing, then editing an existing Post created on another device) seemed to work fine.
> 
> 
> All of this using a Google Pixel 4a.
> 
> I also verified most of the changes in this PR are due to new build errors that appear from changing the `targetSdk` value to 30, so that's fine.
> 
> **IMPORTANT**: It is worth noting that when compiling the WordPress-Android app, the stories library will take the `targetSdk` and `compileSdk` from the parent project so, what this really means is that when building WPAndroid if we don't perform the changes [as expressed in this diff](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) then we'd only have tested the **_code_** changes in this PR but **not** the actual SDK change which this PR's main aim is supposed to be cc @ravishanker . I'd only ask that we make a more thorough list of what's been tested and how next time 👍 .
> 
> So, this **needs to be re-tested** once the parent app (WPAndroid) is ready for the change to API level 30, before we update the commit hash of submodule `/libs/stories-android` in the parent repository.
> 
> That said, let's please revert the `<uses-permission>` tag removal, and we can merge afterwards :shipit:
> 
> Thank you!

@mzorz Thank you for reviewing and testing comprehensively. 🙌  Thanks for the assurance

I've reverted the permission.  I removed it for testing and also AS prompting me constantly that it is redundant.

I've been testing with parent WPAndroid app.  I should have pushed that too, to spare you the trouble of having to make them for testing.  I'll push the WPAndroid as well now.

mzorz commented:
> I've reverted the permission. I removed it for testing and also AS prompting me constantly that it is redundant.

> I've been testing with parent WPAndroid app. I should have pushed that too, to spare you the trouble of having to make them for testing. I'll push the WPAndroid as well now

Thank you @ravishanker 👍 - left a comment [there](https://github.com/wordpress-mobile/WordPress-Android/pull/15331#issuecomment-920721336).

In reading more about how this all fits in your migration plan paqN3M-n4-p2, I found that we might need some more [thorough testing](https://developer.android.com/about/versions/11/privacy/storage#scoped-storage) on devices running on Android 11 when targeting SDK 30. For this, TIL there's a tool we can use to force the [flags needed](https://developer.android.com/about/versions/11/privacy/storage#test-scoped-storage). (see [App compatibility Changes](https://developer.android.com/guide/app-compatibility/test-debug) screen in Settings -> System > Advanced > Developer options > App Compatibility Changes.).

With this in mind, I'll run some more tests on this PR before we proceed to merge. If you can do the same, please by all means do 👍 🙇 cc @ParaskP7

mzorz commented:
I've uninstalled the demo (loop) app, re-installed it, changed the flag `FORCE_ENABLED_SCOPED_STORAGE` to `ON` as shown in the screenshot, and was able to use the media picker and navigate the contents both shown right away and also using the device picker (by tapping on the bottom button, then `Choose photo from device` or `Choose video from device`, both worked well).
<img width="299" alt="Screen Shot 2021-09-16 at 11 46 11" src="https://user-images.githubusercontent.com/6597771/133590497-f52dc3dc-db5c-4953-b8bf-5512416ae79a.png">


However, I encountered a crash on WPAndroid, may be unrelated though, but created an issue here https://github.com/wordpress-mobile/WordPress-Android/issues/15333

ravishanker commented:
> Thank you for taking on this task, @ravishanker !
> 
> I've smoke-tested both the demo app and WordPress-Android app with the stories-android library pointing to this branch, and using [this test branch on WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) (as some changes needed to be made in there as well).
> 
>     * camera: taking pictures and recording videos worked as expected
> 
>     * saving: both new pics and videos were composed successfully
> 
>     * sourcing media: using the media picker also worked well (local media, WP media in the WP app, 3rd party app media, remote media).
> 
>     * integration with Gutenberg (creating a post from zero, then editing, then editing an existing Post created on another device) seemed to work fine.
> 
> 
> All of this using a Google Pixel 4a.
> 
> I also verified most of the changes in this PR are due to new build errors that appear from changing the `targetSdk` value to 30, so that's fine.
> 
> **IMPORTANT**: It is worth noting that when compiling the WordPress-Android app, the stories library will take the `targetSdk` and `compileSdk` from the parent project so, what this really means is that when building WPAndroid if we don't perform the changes [as expressed in this diff](https://github.com/wordpress-mobile/WordPress-Android/compare/mzorz/test-sdk30-stories?expand=1) then we'd only have tested the **_code_** changes in this PR but **not** the actual SDK change which this PR's main aim is supposed to be cc @ravishanker . I'd only ask that we make a more thorough list of what's been tested and how next time 👍 .
> 
> So, this **needs to be re-tested** once the parent app (WPAndroid) is ready for the change to API level 30, before we update the commit hash of submodule `/libs/stories-android` in the parent repository.
> 
> That said, let's please revert the `<uses-permission>` tag removal, and we can merge afterwards :shipit:
> 
> Thank you!

@mzorz I'm attempting to complete upgrade task.  I've already addressed all the changes suggested.  Are you able to remove this, to help me proceed with merger.  Thank you 🙏

mzorz commented:
Could you update this branch with `develop`? Also I can see the related [WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/15331) has [conflicts](https://github.com/wordpress-mobile/WordPress-Android/pull/15331/conflicts), should these be fixed before we do a final round of testing and merge please? 🙏 @ravishanker 

Also, a PR has just been merged for [composite builds](https://github.com/Automattic/stories-android/pull/713). Do we know how this update to SDK 30 may play together? (I haven't thought of it through deeply, so summoning here @oguzkocer just to double check whether we have a clear path to merge develop into this branch, test, then merge the PR or we need something else at this point).

oguzkocer commented:
Thanks for the ping @mzorz. As long as this PR and WPAndroid counterpart are updated with latest `develop` (and merge conflicts are resolved), we should be all good.

ravishanker committed:
Merge branch 'develop' into fix/15322-upgrade-to-android-11-api-level-30

mzorz reviewed (approved):
Thank you for these changes @ravishanker ! 🙌 
LGTM :shipit:  - I'll merge and create a new release tag on the repo that can be referenced by WPAndroid
